File: android/dynamics-migration-tool/templates/sql/room-bridge/GDRoomDatabase.java
// [BB_DYNAMICS-MIGRATION] This file was added as part of BlackBerry Dynamics migration.
// Provenance: kit-authored. License: Apache-2.0 — see templates/_LICENSE-NOTICE.md.

package __APP_PACKAGE__;

import android.content.ContentValues;
import android.database.Cursor;
import android.os.CancellationSignal;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.sqlite.db.SupportSQLiteQuery;
import androidx.sqlite.db.SupportSQLiteStatement;

// [BB_DYNAMICS-MIGRATION] Dynamics SQLiteDatabase wraps the secure container's database.
import com.good.gd.database.sqlite.SQLiteDatabase;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

/**
 * SupportSQLiteDatabase adapter that wraps a Dynamics SQLiteDatabase.
 *
 * Room calls this interface for all SQL operations. Every method delegates to
 * the underlying Dynamics db, which writes encrypted data inside the container.
 *
 * IMPORTANT — close() must declare throws IOException (from java.io.Closeable).
 * Do NOT change it to throws Exception — the class will fail to compile.
 */
// [BB_DYNAMICS-MIGRATION] Implements Room's SupportSQLiteDatabase wrapping Dynamics SQLiteDatabase.
final class GDRoomDatabase implements SupportSQLiteDatabase {

    // [BB_DYNAMICS-MIGRATION] The Dynamics secure database — all I/O is encrypted by the container.
    private final SQLiteDatabase db;

    GDRoomDatabase(@NonNull SQLiteDatabase db) {
        this.db = db;
    }

    // ---- Queries ------------------------------------------------------------

    @Override
    public Cursor query(@NonNull String sql) {
        return db.rawQuery(sql, null);
    }

    @Override
    public Cursor query(@NonNull String sql, @NonNull Object[] bindArgs) {
        String[] strArgs = toStringArray(bindArgs);
        return db.rawQuery(sql, strArgs);
    }

    @Override
    public Cursor query(@NonNull SupportSQLiteQuery query) {
        GDRoomQueryBindingHelper helper = new GDRoomQueryBindingHelper();
        query.bindTo(helper);
        return db.rawQuery(query.getSql(), helper.getBindings());
    }

    @Override
    public Cursor query(@NonNull SupportSQLiteQuery query, @Nullable CancellationSignal cancellationSignal) {
        // CancellationSignal not supported by Dynamics SQLite — delegate without it.
        return query(query);
    }

    // ---- DML ----------------------------------------------------------------

    @Override
    public long insert(@NonNull String table, int conflictAlgorithm, @NonNull ContentValues values) {
        return db.insertWithOnConflict(table, null, values, conflictAlgorithm);
    }

    @Override
    public int delete(@NonNull String table, @Nullable String whereClause, @Nullable Object[] whereArgs) {
        return db.delete(table, whereClause, toStringArray(whereArgs));
    }

    @Override
    public int update(@NonNull String table, int conflictAlgorithm,
                      @NonNull ContentValues values,
                      @Nullable String whereClause, @Nullable Object[] whereArgs) {
        return db.updateWithOnConflict(table, values, whereClause, toStringArray(whereArgs), conflictAlgorithm);
    }

    @Override
    public void execSQL(@NonNull String sql) {
        db.execSQL(sql);
    }

    @Override
    public void execSQL(@NonNull String sql, @NonNull Object[] bindArgs) {
        db.execSQL(sql, bindArgs);
    }

    // ---- Compiled statements ------------------------------------------------

    @NonNull
    @Override
    public SupportSQLiteStatement compileStatement(@NonNull String sql) {
        return new GDRoomStatement(db.compileStatement(sql));
    }

    // ---- Transactions -------------------------------------------------------

    @Override
    public void beginTransaction() {
        db.beginTransaction();
    }

    @Override
    public void beginTransactionNonExclusive() {
        db.beginTransactionNonExclusive();
    }

    @Override
    public void beginTransactionWithListener(
            @NonNull android.database.sqlite.SQLiteTransactionListener listener) {
        // [BB_DYNAMICS-MIGRATION] Android SQLiteTransactionListener must be wrapped to
        // satisfy Dynamics SQLiteDatabase.beginTransactionWithListener() parameter type.
        db.beginTransactionWithListener(new com.good.gd.database.sqlite.SQLiteTransactionListener() {
            @Override public void onBegin()    { listener.onBegin(); }
            @Override public void onCommit()   { listener.onCommit(); }
            @Override public void onRollback() { listener.onRollback(); }
        });
    }

    @Override
    public void beginTransactionWithListenerNonExclusive(
            @NonNull android.database.sqlite.SQLiteTransactionListener listener) {
        db.beginTransactionWithListenerNonExclusive(
                new com.good.gd.database.sqlite.SQLiteTransactionListener() {
                    @Override public void onBegin()    { listener.onBegin(); }
                    @Override public void onCommit()   { listener.onCommit(); }
                    @Override public void onRollback() { listener.onRollback(); }
                });
    }

    @Override
    public void endTransaction() {
        db.endTransaction();
    }

    @Override
    public void setTransactionSuccessful() {
        db.setTransactionSuccessful();
    }

    @Override
    public boolean inTransaction() {
        return db.inTransaction();
    }

    // ---- State --------------------------------------------------------------

    @Override
    public boolean isDbLockedByCurrentThread() {
        return db.isDbLockedByCurrentThread();
    }

    @Override
    public boolean yieldIfContendedSafely() {
        return db.yieldIfContendedSafely();
    }

    @Override
    public boolean yieldIfContendedSafely(long sleepAfterYieldDelay) {
        return db.yieldIfContendedSafely(sleepAfterYieldDelay);
    }

    @Override
    public int getVersion() {
        return db.getVersion();
    }

    @Override
    public void setVersion(int version) {
        db.setVersion(version);
    }

    @Override
    public long getMaximumSize() {
        return db.getMaximumSize();
    }

    @Override
    public long setMaximumSize(long numBytes) {
        return db.setMaximumSize(numBytes);
    }

    @Override
    public long getPageSize() {
        return db.getPageSize();
    }

    @Override
    public void setPageSize(long numBytes) {
        db.setPageSize(numBytes);
    }

    @Nullable
    @Override
    public String getPath() {
        return db.getPath();
    }

    @Override
    public boolean isOpen() {
        return db.isOpen();
    }

    @Override
    public boolean needUpgrade(int newVersion) {
        return db.needUpgrade(newVersion);
    }

    @Override
    public boolean isReadOnly() {
        return db.isReadOnly();
    }

    @Override
    public boolean isWriteAheadLoggingEnabled() {
        return db.isWriteAheadLoggingEnabled();
    }

    @Override
    public void setForeignKeyConstraintsEnabled(boolean enable) {
        db.setForeignKeyConstraintsEnabled(enable);
    }

    @Override
    public boolean enableWriteAheadLogging() {
        return db.enableWriteAheadLogging();
    }

    @Override
    public void disableWriteAheadLogging() {
        db.disableWriteAheadLogging();
    }

    @Nullable
    @Override
    public List<Pair<String, String>> getAttachedDbs() {
        return db.getAttachedDbs();
    }

    @Override
    public boolean isDatabaseIntegrityOk() {
        return db.isDatabaseIntegrityOk();
    }

    // Commonly missed — from SupportSQLiteDatabase interface.
    @Override
    public void setMaxSqlCacheSize(int cacheSize) {
        db.setMaxSqlCacheSize(cacheSize);
    }

    @Override
    public void setLocale(@NonNull Locale locale) {
        db.setLocale(locale);
    }

    // IMPORTANT: close() MUST declare throws IOException (from java.io.Closeable).
    // Declaring throws Exception will NOT satisfy the interface and causes a compile error.
    @Override
    public void close() throws IOException {
        db.close();
    }

    // ---- Helpers ------------------------------------------------------------

    @Nullable
    private static String[] toStringArray(@Nullable Object[] args) {
        if (args == null) return null;
        String[] result = new String[args.length];
        for (int i = 0; i < args.length; i++) {
            if (args[i] instanceof byte[]) {
                throw new UnsupportedOperationException(
                        "BLOB argument at index " + i
                                + " cannot be passed through rawQuery/delete/update String[] bindings. "
                                + "Use a compiled SQLiteStatement with bindBlob() for BLOB predicates."
                );
            }
            result[i] = args[i] == null ? null : String.valueOf(args[i]);
        }
        return result;
    }
}

File: android/dynamics-migration-tool/templates/sql/room-bridge/GDRoomDatabase.kt
// [BB_DYNAMICS-MIGRATION] This file was added as part of BlackBerry Dynamics migration.
// Provenance: kit-authored. License: Apache-2.0 — see templates/_LICENSE-NOTICE.md.

package __APP_PACKAGE__

import android.content.ContentValues
import android.database.Cursor
import android.os.CancellationSignal
import android.util.Pair
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteQuery
import androidx.sqlite.db.SupportSQLiteStatement
// [BB_DYNAMICS-MIGRATION] Dynamics SQLiteDatabase — all data encrypted by the secure container.
import com.good.gd.database.sqlite.SQLiteDatabase
import java.io.IOException
import java.util.Locale

// [BB_DYNAMICS-MIGRATION] Implements Room's SupportSQLiteDatabase wrapping Dynamics SQLiteDatabase.
internal class GDRoomDatabase(private val db: SQLiteDatabase) : SupportSQLiteDatabase {

    override fun query(sql: String): Cursor = db.rawQuery(sql, null)

    override fun query(sql: String, bindArgs: Array<out Any?>): Cursor =
        db.rawQuery(sql, bindArgs.toStringArray())

    override fun query(query: SupportSQLiteQuery): Cursor {
        val helper = GDRoomQueryBindingHelper()
        query.bindTo(helper)
        return db.rawQuery(query.sql, helper.getBindings())
    }

    override fun query(query: SupportSQLiteQuery, cancellationSignal: CancellationSignal?): Cursor =
        query(query)

    override fun insert(table: String, conflictAlgorithm: Int, values: ContentValues): Long =
        db.insertWithOnConflict(table, null, values, conflictAlgorithm)

    override fun delete(table: String, whereClause: String?, whereArgs: Array<out Any?>?): Int =
        db.delete(table, whereClause, whereArgs?.toStringArray())

    override fun update(
        table: String, conflictAlgorithm: Int, values: ContentValues,
        whereClause: String?, whereArgs: Array<out Any?>?
    ): Int = db.updateWithOnConflict(table, values, whereClause, whereArgs?.toStringArray(), conflictAlgorithm)

    override fun execSQL(sql: String) = db.execSQL(sql)

    override fun execSQL(sql: String, bindArgs: Array<out Any?>) = db.execSQL(sql, bindArgs)

    override fun compileStatement(sql: String): SupportSQLiteStatement =
        GDRoomStatement(db.compileStatement(sql))

    override fun beginTransaction() = db.beginTransaction()
    override fun beginTransactionNonExclusive() = db.beginTransactionNonExclusive()

    override fun beginTransactionWithListener(
        listener: android.database.sqlite.SQLiteTransactionListener
    ) {
        // [BB_DYNAMICS-MIGRATION] Wrap Android SQLiteTransactionListener for Dynamics parameter type.
        db.beginTransactionWithListener(object : com.good.gd.database.sqlite.SQLiteTransactionListener {
            override fun onBegin() = listener.onBegin()
            override fun onCommit() = listener.onCommit()
            override fun onRollback() = listener.onRollback()
        })
    }

    override fun beginTransactionWithListenerNonExclusive(
        listener: android.database.sqlite.SQLiteTransactionListener
    ) {
        db.beginTransactionWithListenerNonExclusive(
            object : com.good.gd.database.sqlite.SQLiteTransactionListener {
                override fun onBegin() = listener.onBegin()
                override fun onCommit() = listener.onCommit()
                override fun onRollback() = listener.onRollback()
            })
    }

    override fun endTransaction() = db.endTransaction()
    override fun setTransactionSuccessful() = db.setTransactionSuccessful()
    override fun inTransaction(): Boolean = db.inTransaction()
    override fun isDbLockedByCurrentThread(): Boolean = db.isDbLockedByCurrentThread
    override fun yieldIfContendedSafely(): Boolean = db.yieldIfContendedSafely()
    override fun yieldIfContendedSafely(sleepAfterYieldDelay: Long): Boolean =
        db.yieldIfContendedSafely(sleepAfterYieldDelay)

    override fun getVersion(): Int = db.version
    override fun setVersion(version: Int) { db.version = version }
    override fun getMaximumSize(): Long = db.maximumSize
    override fun setMaximumSize(numBytes: Long): Long = db.setMaximumSize(numBytes)
    override fun getPageSize(): Long = db.pageSize
    override fun setPageSize(numBytes: Long) { db.pageSize = numBytes }
    override fun getPath(): String? = db.path
    override fun isOpen(): Boolean = db.isOpen
    override fun needUpgrade(newVersion: Int): Boolean = db.needUpgrade(newVersion)
    override fun isReadOnly(): Boolean = db.isReadOnly
    override fun isWriteAheadLoggingEnabled(): Boolean = db.isWriteAheadLoggingEnabled
    override fun setForeignKeyConstraintsEnabled(enable: Boolean) =
        db.setForeignKeyConstraintsEnabled(enable)

    override fun enableWriteAheadLogging(): Boolean = db.enableWriteAheadLogging()
    override fun disableWriteAheadLogging() = db.disableWriteAheadLogging()
    override fun getAttachedDbs(): MutableList<Pair<String, String>>? = db.attachedDbs
    override fun isDatabaseIntegrityOk(): Boolean = db.isDatabaseIntegrityOk

    // Commonly missed — setMaxSqlCacheSize and setLocale.
    override fun setMaxSqlCacheSize(cacheSize: Int) = db.setMaxSqlCacheSize(cacheSize)
    override fun setLocale(locale: Locale) = db.setLocale(locale)

    // IMPORTANT: close() throws IOException (from Closeable). NOT throws Exception.
    @Throws(IOException::class)
    override fun close() = db.close()

    private fun Array<out Any?>.toStringArray(): Array<String?> =
        Array(size) { i ->
            val value = this[i]
            if (value is ByteArray) {
                throw UnsupportedOperationException(
                    "BLOB argument at index $i cannot be passed through " +
                        "rawQuery/delete/update String[] bindings. " +
                        "Use a compiled SQLiteStatement with bindBlob() for BLOB predicates."
                )
            }
            value?.toString()
        }
}
